nvidia_gpu_stats/src/main.rs

```diff
@@ -0,0 +1,356 @@
+    interval: f64,
+}
+
+/// Struct to hold GPU metrics. Metrics are stored in a BTreeMap to ensure
+/// consistent ordering of keys in the output JSON.
+/// The output map is flat to make it easier to parse in downstream applications.
+/// cuda_version: The version of CUDA installed on the system.
+/// gpu.count: The total number of GPUs detected in the system.
+/// gpu.{i}.name: The name of the GPU at index i (e.g., Tesla T4).
+/// gpu.{i}.brand: The brand of the GPU at index i (e.g., GeForce, Nvidia).
+/// gpu.{i}.fanSpeed: The current fan speed of the GPU at index i (in percentage).
+/// gpu.{i}.encoderUtilization: The utilization of the GPU's encoder at index i (in percentage).
+/// gpu.{i}.gpu: The overall GPU utilization at index i (in percentage).
+/// gpu.{i}.memory: The GPU memory utilization at index i (in percentage).
+/// gpu.{i}.memoryTotal: The total memory of the GPU at index i (in bytes).
+/// gpu.{i}.memoryAllocated: The percentage of GPU memory allocated at index i.
+/// gpu.{i}.memoryAllocatedBytes: The amount of GPU memory allocated at index i (in bytes).
+/// gpu.{i}.temp: The temperature of the GPU at index i (in Celsius).
+/// gpu.{i}.powerWatts: The power consumption of the GPU at index i (in Watts).
+/// gpu.{i}.enforcedPowerLimitWatts: The enforced power limit of the GPU at index i (in Watts).
+/// gpu.{i}.powerPercent: The percentage of power limit being used by the GPU at index i.
+/// gpu.{i}.graphicsClock: The current graphics clock speed of the GPU at index i (in MHz).
+/// gpu.{i}.memoryClock: The current memory clock speed of the GPU at index i (in MHz).
+/// gpu.{i}.pcieLinkGen: The current PCIe link generation of the GPU at index i.
+/// gpu.{i}.pcieLinkSpeed: The current PCIe link speed of the GPU at index i (in bits per second).
+/// gpu.{i}.pcieLinkWidth: The current PCIe link width of the GPU at index i.
+/// gpu.{i}.maxPcieLinkGen: The maximum PCIe link generation supported by the GPU at index i.
+/// gpu.{i}.maxPcieLinkWidth: The maximum PCIe link width supported by the GPU at index i.
+/// gpu.{i}.cudaCores: The number of CUDA cores in the GPU at index i.
+/// gpu.{i}.architecture: The architecture of the GPU at index i (e.g., Ampere, Turing).
+/// gpu.process.{i}.*: Various metrics specific to the monitored process
+///    (if the GPU is in use by the process). These include GPU utilization, memory utilization,
+///     temperature, and power consumption.
+/// _timestamp: The Unix timestamp when the metrics were collected.
+///
+/// Note that {i} represents the index of each GPU in the system, starting from 0.
+#[derive(Serialize)]
+struct GpuMetrics {
+    #[serde(flatten)]
+    metrics: BTreeMap<String, serde_json::Value>,
+}
+
+/// Function to get child process IDs for a given parent PID
+fn get_child_pids(pid: i32) -> Vec<i32> {
+    let output = Command::new("pgrep")
+        .args(&["-P", &pid.to_string()])
+        .output()
+        .expect("Failed to execute pgrep");
```

Comment:
This can panic. It'd be better to change the return value to a `Result<...>` and let the caller handle the error case.

Reply:
yea, this was left from a previous iteration, I'm replacing this with something more language-native.